File: Quices/quiz_23_02_26/afd_quiz.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>

// Estados del AFD (Q). Los nombres son los que explicas en la exposición.
typedef enum {
    ESTADO_INICIAL,
    ESTADO_ID_ACEPTADO,
    ESTADO_ID_ESPERA_DIGITO,
    ESTADO_SUMA,
    ESTADO_INCREMENTO,
    ESTADO_ENTERO,
    ESTADO_RECHAZO,
    CANTIDAD_ESTADOS
} Estado;

// Símbolos del alfabeto Σ (en categorías, no char por char).
typedef enum {
    S_UPPER,
    S_LOWER,
    S_DIGIT,
    S_PLUS,
    S_OTHER,
    S_COUNT
} Simbolo;

// δ (delta): función de transición del AFD.
// DELTA[estado_actual][simbolo] = estado_siguiente
static const Estado DELTA[CANTIDAD_ESTADOS][S_COUNT] = {
    [ESTADO_INICIAL] = {
        [S_UPPER] = ESTADO_ID_ACEPTADO,
        [S_LOWER] = ESTADO_ID_ACEPTADO,
        [S_DIGIT] = ESTADO_ENTERO,
        [S_PLUS] = ESTADO_SUMA,
        [S_OTHER] = ESTADO_RECHAZO
    },
    [ESTADO_ID_ACEPTADO] = {
        [S_UPPER] = ESTADO_RECHAZO,
        [S_LOWER] = ESTADO_ID_ESPERA_DIGITO,
        [S_DIGIT] = ESTADO_RECHAZO,
        [S_PLUS] = ESTADO_RECHAZO,
        [S_OTHER] = ESTADO_RECHAZO
    },
    [ESTADO_ID_ESPERA_DIGITO] = {
        [S_UPPER] = ESTADO_RECHAZO,
        [S_LOWER] = ESTADO_RECHAZO,
        [S_DIGIT] = ESTADO_ID_ACEPTADO,
        [S_PLUS] = ESTADO_RECHAZO,
        [S_OTHER] = ESTADO_RECHAZO
    },
    [ESTADO_SUMA] = {
        [S_UPPER] = ESTADO_RECHAZO,
        [S_LOWER] = ESTADO_RECHAZO,
        [S_DIGIT] = ESTADO_RECHAZO,
        [S_PLUS] = ESTADO_INCREMENTO,
        [S_OTHER] = ESTADO_RECHAZO
    },
    [ESTADO_INCREMENTO] = {
        [S_UPPER] = ESTADO_RECHAZO,
        [S_LOWER] = ESTADO_RECHAZO,
        [S_DIGIT] = ESTADO_RECHAZO,
        [S_PLUS] = ESTADO_RECHAZO,
        [S_OTHER] = ESTADO_RECHAZO
    },
    [ESTADO_ENTERO] = {
        [S_UPPER] = ESTADO_RECHAZO,
        [S_LOWER] = ESTADO_RECHAZO,
        [S_DIGIT] = ESTADO_RECHAZO,
        [S_PLUS] = ESTADO_RECHAZO,
        [S_OTHER] = ESTADO_RECHAZO
    },
    [ESTADO_RECHAZO] = {
        [S_UPPER] = ESTADO_RECHAZO,
        [S_LOWER] = ESTADO_RECHAZO,
        [S_DIGIT] = ESTADO_RECHAZO,
        [S_PLUS] = ESTADO_RECHAZO,
        [S_OTHER] = ESTADO_RECHAZO
    }
};

// F: estados finales. Si termino en uno de estos, la cadena se acepta.
static int es_estado_final(Estado estado) {
    return estado == ESTADO_ID_ACEPTADO || estado == ESTADO_SUMA || estado == ESTADO_INCREMENTO || estado == ESTADO_ENTERO;
}

// Esta función clasifica un char real en una categoría del alfabeto.
// O sea, esto es pa que el autómata trabaje por clases de símbolo.
static Simbolo clasificar(unsigned char c) {
    if (c >= 'A' && c <= 'Z') {
        return S_UPPER;
    }
    if (c >= 'a' && c <= 'z') {
        return S_LOWER;
    }
    if (c >= '0' && c <= '9') {
        return S_DIGIT;
    }
    if (c == '+') {
        return S_PLUS;
    }
    return S_OTHER;
}

// Corre el AFD completo sobre una cadena.
int acepta(const char *cadena) {
    // Arrancamos en q0.
    Estado estado = ESTADO_INICIAL;
    // Recorremos la cadena símbolo por símbolo aplicando δ.
    for (size_t i = 0; cadena[i] != '\0'; i++) {
        Simbolo simbolo = clasificar((unsigned char)cadena[i]);
        estado = DELTA[estado][simbolo];
    }
    // Si el último estado es final, acepta.
    return es_estado_final(estado);
}

int main(int argc, char *argv[]) {
    // Ruta por argumento o archivo.txt por defecto.
    const char *ruta = (argc > 1) ? argv[1] : "archivo.txt";
    FILE *archivo = fopen(ruta, "r");
    char cadena[1024];
    // Bandera pa saber si el archivo traía al menos una línea.
    int hay_lineas = 0;

    if (archivo == NULL) {
        // Si no abre, devolvemos lo que pide el enunciado.
        printf("No Acepta\n");
        return 0;
    }

    // Recorremos TODO el archivo línea por línea.
    while (fgets(cadena, sizeof(cadena), archivo) != NULL) {
        hay_lineas = 1;
        // Quitamos solo salto de línea final.
        cadena[strcspn(cadena, "\r\n")] = '\0';
        if (acepta(cadena)) {
            printf("Acepta\n");
        } else {
            printf("No Acepta\n");
        }
    }

    fclose(archivo);

    // Archivo vacío => no hay nada aceptable.
    if (!hay_lineas) {
        printf("No Acepta\n");
    }

    return 0;
}

File: Quices/quiz_23_02_26/afd_quiz.py
import sys
import os


# Estos son los tipos de símbolo que usamos pa simplificar el alfabeto.
# En vez de guardar cada char posible, lo metemos en una categoría.
SIMBOLOS = ("UPPER", "LOWER", "DIGIT", "PLUS", "OTHER")

# M = (Q, Σ, δ, q0, F)
# Q: conjunto de estados del AFD.
Q = (
    "q_inicial",
    "q_id_aceptado",
    "q_id_espera_digito",
    "q_suma",
    "q_incremento",
    "q_entero",
    "q_rechazo",
)
# Σ: alfabeto (acá representado por categorías de símbolo).
SIGMA = set(SIMBOLOS)
# q0: estado inicial.
Q0 = "q_inicial"
# F: estados de aceptación (si termino aquí, la cadena se acepta).
F = {"q_id_aceptado", "q_suma", "q_incremento", "q_entero"}

# δ: función de transición.
# Esta tabla dice: "si estoy en tal estado y leo tal símbolo, me voy a tal estado.
DELTA = {
    "q_inicial": {
        "UPPER": "q_id_aceptado",
        "LOWER": "q_id_aceptado",
        "DIGIT": "q_entero",
        "PLUS": "q_suma",
        "OTHER": "q_rechazo",
    },
    "q_id_aceptado": {
        "UPPER": "q_rechazo",
        "LOWER": "q_id_espera_digito",
        "DIGIT": "q_rechazo",
        "PLUS": "q_rechazo",
        "OTHER": "q_rechazo",
    },
    "q_id_espera_digito": {
        "UPPER": "q_rechazo",
        "LOWER": "q_rechazo",
        "DIGIT": "q_id_aceptado",
        "PLUS": "q_rechazo",
        "OTHER": "q_rechazo",
    },
    "q_suma": {
        "UPPER": "q_rechazo",
        "LOWER": "q_rechazo",
        "DIGIT": "q_rechazo",
        "PLUS": "q_incremento",
        "OTHER": "q_rechazo",
    },
    "q_incremento": {
        "UPPER": "q_rechazo",
        "LOWER": "q_rechazo",
        "DIGIT": "q_rechazo",
        "PLUS": "q_rechazo",
        "OTHER": "q_rechazo",
    },
    "q_entero": {
        "UPPER": "q_rechazo",
        "LOWER": "q_rechazo",
        "DIGIT": "q_rechazo",
        "PLUS": "q_rechazo",
        "OTHER": "q_rechazo",
    },
    "q_rechazo": {
        "UPPER": "q_rechazo",
        "LOWER": "q_rechazo",
        "DIGIT": "q_rechazo",
        "PLUS": "q_rechazo",
        "OTHER": "q_rechazo",
    },
}


def clasificar(caracter: str) -> str:
    # Esta función traduce un char real a una categoría del alfabeto.
    # O sea, esto es pa que el AFD no dependa de chars sueltos sino de clases.
    if "A" <= caracter <= "Z":
        return "UPPER"
    if "a" <= caracter <= "z":
        return "LOWER"
    if "0" <= caracter <= "9":
        return "DIGIT"
    if caracter == "+":
        return "PLUS"
    return "OTHER"


def acepta(cadena: str) -> bool:
    # Arrancamos siempre en q0 (estado inicial).
    estado = Q0
    # Recorremos la cadena char por char, aplicando δ cada vez.
    for caracter in cadena:
        simbolo = clasificar(caracter)
        estado = DELTA[estado][simbolo]
    # Si al final quedé en un estado final, acepta; si no, rechaza.
    return estado in F


def main() -> None:
    # Si me pasan ruta por consola, la uso; si no, uso archivo.txt por defecto.
    ruta = sys.argv[1] if len(sys.argv) > 1 else "archivo.txt"
    # Esto es pa no romperse si me pasan un nombre que no existe.
    if not os.path.exists(ruta):
        ruta = "archivo.txt"

    try:
        # Leemos TODO el archivo, línea por línea.
        with open(ruta, "r", encoding="utf-8") as archivo:
            hay_lineas = False
            for linea in archivo:
                hay_lineas = True
                # Solo quitamos salto de línea, no tocamos el contenido interno.
                cadena = linea.rstrip("\r\n")
                # Para cada línea imprimimos exactamente lo pedido.
                print("Acepta" if acepta(cadena) else "No Acepta")
    except OSError:
        # Si no se puede abrir el archivo, lo tomamos como no aceptado.
        print("No Acepta")
        return

    # Si el archivo está vacío, también reportamos no aceptación. por q no quiro hacer un try catch para eso, y es más fácil hacer esto que manejar el error de EOF.
    if not hay_lineas:
        print("No Acepta")


if __name__ == "__main__":
    main()
